56221 rest install plugin from url

BrianHenryIE opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/56221

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

BrianHenryIE committed:
Add test case to install link-manager from URL

BrianHenryIE committed:
Merge branch 'trunk' into 56221-rest-install-plugin-from-url

BrianHenryIE committed:
Fix ticket number on test case

BrianHenryIE committed:
Accept remote URLs for plugin install.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-plugins-controller.php:
```diff
@@ -51,10 +51,16 @@ public function register_routes() {
+							'type'        => 'string',
+							'required'    => false,
+							'description' => __( 'A URL to a zip or tar archive.' ),
+							'pattern'     => '^https?://[^\/]*/.*(zip|tar|tar\.gz)$',
```
Instead of using a `pattern`, this should use the `format` keyword set to `uri`.

BrianHenryIE replied:
Can it use both? i.e. use `format` to check is it a URL, and use a pattern to check it's an allowed archive format.

TimothyBJacobs replied:
Yes, it's additive, so all keywords provided will be evaluated. I'd then remove the URL validation portion, and just focus on it ending with one of the expected extensions.

BrianHenryIE replied:
OK. I consider this part resolved: https://www.phpliveregex.com/p/GAT

BrianHenryIE replied:
Although I set `'required'    => false,` on the two params. Can I leave that out? I think it's clearer to leave it out, rather than suggesting it fully is required or is not required.

TimothyBJacobs replied:
Correct, the keyword can be omitted to treat it as a non-required field.

BrianHenryIE replied:
Done.

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-plugins-controller.php:
```diff
@@ -380,25 +392,27 @@ public function create_item( $request ) {
-			},
-			$api->language_packs
-		);
+		if ( isset( $api ) ) {
```
I would instead initialize `$language_packs` to an empty array before the branch at the top of the function. Then you can set it's value in the same slug branch after we've determined the response isn't a `WP_Error`.

BrianHenryIE replied:
Yeah, that's nicer.
https://github.com/WordPress/wordpress-develop/pull/3515/commits/f2bec388bfa2f8f039f9c12b4db6893fc3783de6

BrianHenryIE replied:
While adding the validate_callback, I realised a valid case where both slug and url are present is for URLs pointing to wordpress.org. E.g. you can install an older version of a plugin by using the URL, and still want the language pack installed from the repo.

BrianHenryIE committed:
initialize $language_packs within branches of create_item

BrianHenryIE committed:
Validate create plugin url parameter is archive

Validate using UIR format, plus regex to ensure it is an archive file
https://www.phpliveregex.com/p/GAT

BrianHenryIE committed:
Add create validate_callback at the route-level

If both url and slug are present, and the URL is a wordpress.org download URL, the language pack will be downloaded too.

BrianHenryIE committed:
Remove `required=>true|false` on xor required args

One of either `slug` or `url` is required, so setting both to true is incorrect, but setting both to false also miscommunicates the requirement. Omitting `required` seems like the best option here.

dd32 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-plugins-controller.php:
```diff
@@ -278,39 +284,56 @@ public function create_item( $request ) {
-				),
-			)
-		);
+			// Verify filesystem is accessible first.
+			$filesystem_available = $this->is_filesystem_available();
+			if ( is_wp_error( $filesystem_available ) ) {
+				return $filesystem_available;
+			}
```
This should be before the isset() checks. ie. It should execute even if the `url` parameter is used.
